litex/build/quicklogic/f4pga.py

```diff
@@ -47,48 +48,59 @@ def build_io_constraints(self):
         return (self._build_name + ".pcf", "PCF")
 
-    # Build Makefile -------------------------------------------------------------------------------
+    # Timing constraints (.sdc) --------------------------------------------------------------------
+
+    def build_timing_constraints(self, vns):
+        sdc = []
+        for clk, [period, name] in sorted(self.clocks.items(), key=lambda x: x[0].duid):
+            clk_sig = vns.get_name(clk)
+            sdc.append("create_clock -period {} {}".format(str(period), clk_sig))
+        tools.write_to_file(self._build_name + ".sdc", "\n".join(sdc))
+        return (self._build_name + ".sdc", "SDC")
+
+    # Build flow.json ------------------------------------------------------------------------------
 
     def build_script(self):
-        makefile = []
-
-        # Define Paths.
-        makefile.append("mkfile_path := $(abspath $(lastword $(MAKEFILE_LIST)))")
-        makefile.append("current_dir := $(patsubst %/,%,$(dir $(mkfile_path)))")
-        # bit -> h and bit -> bin requires TOP_F
-        makefile.append(f"TOP_F={self._build_name}")
-
-        # Create Project.
-        # FIXME: Only use top file for now and ignore .init files.
-        makefile.append("all: {top}_bit.h {top}.bin build/{top}.bit".format(top=self._build_name))
-        # build bit file (default)
-        makefile.append(f"build/{self._build_name}.bit:")
-        makefile.append("\tql_symbiflow -compile -d {device} -P {part} -v {verilog} -t {top} -p {pcf}".format(
-            device  = self.platform.device,
-            part    = {"ql-eos-s3": "PU64"}.get(self.platform.device),
-            verilog = f"{self._build_name}.v",
-            top     = self._build_name,
-            pcf     = f"{self._build_name}.pcf"
-        ))
-        # build header to include in CPU firmware
-        makefile.append("{top}_bit.h: build/{top}.bit".format(top=self._build_name))
-        makefile.append(f"\t(cd build; TOP_F=$(TOP_F) symbiflow_write_bitheader)")
-        # build binary to write in dedicated FLASH area
-        makefile.append("{top}.bin: build/{top}.bit".format(top=self._build_name))
-        makefile.append(f"\t(cd build; TOP_F=$(TOP_F) symbiflow_write_binary)")
-
-        # Generate Makefile.
-        tools.write_to_file("Makefile", "\n".join(makefile))
-
-        return "Makefile"
+        part = {"ql-eos-s3": "PU64"}.get(self.platform.device)
+        flow = {
+            "default_part": "EOS3FF512-PDN64",
+            "values": {
+                "top": self._build_name
+            },
+            "dependencies": {
+                "sources": [
+                    f"{self._build_name}.v"
+                ],
+                "synth_log": "synth.log",
+                "pack_log": "pack.log",
+                "analysis_log": "analysis.log"
+            },
+            "EOS3FF512-PDN64": {
+                "default_target": "bitstream",
+                "dependencies": {
+                    "build_dir": "build",
```

Comment:
By setting `build_dir` to *build* directory all output files are located in this sub-dir instead of directly into *gateware* (as for all toolchains). But by using `self._build_dir` instead of build `prepare_cmd` will override *.sdc* file: what do you think about renaming *xxx.sdc* (in `build_timing_constraints`) by *xxx_in.sdc* ?